### PR TITLE
Fixed autoscroll being stuck when already scrolled to top/bottom

### PR DESCRIPTION
**Pull request by hsource**
## Motivation

1. When trying to scroll down by dragging an item while the list is scrolled at the top, the scroll never starts
2. When trying to build the app, we get the error:

```
node_modules/react-native-reanimated/react-native-reanimated.d.ts:477:7 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

477   } = Animated;
          ~~~~~~~~


Found 1 error.
```

## Debugging for scrolling issue

This was happening because:

1. `autoscroll` is called, and it calls `scrollToAsync`, which creates a promise
2. `onScroll` is called even before the scroll happens, and immediately calls `this.onAutoscrollComplete` with the current scroll position (at the top of the document) because `this.isScrolledUp` is true. This then resolves the `scrollToAsync` promise with the current document state (not scrolled, still at the top)
3. `autoscroll` calls itself recursively, but with `nextScrollParams` being the same as the first set of `nextScrollParams`, which is at the top of the document. It then tries to scroll to the same position

This makes the document be constantly trying to scroll to the same position, but not starting

## Fix for scrolling issue

1. Ensure `autoscroll` can't be called simultaneously so that it never tries to repeatedly scroll to the same place
   - It's now only called in the pan handler, and a variable ensures that only one instance is running
   - I made it a loop instead of recursive to make this easier
2. Instead of only checking `this.isScrolledUp` and `this.isScrolledDown` and marking the scroll as done immediately, we should check only `this.isScrolledUp` only while scrolling up and `this.isScrolledDown` only while scrolling down

## Fix for typings

- Upgraded react-native-reanimated types to fix the issue with latest Typescript
- Fixed our typings for the new reanimated typings

## Testing

| Before | After |
| --- | --- |
| Before | After |

### CODE REVIEW

**Comment by hsource** [comment]
@computerjazz - can you take a look? Thanks for this library! It works really well!

Also a note unrelated to this change - I don't think you actually need to build the Typescript when releasing the package, as the Metro bundler used by React Native compiles Typescript in its default configuration anyways.

File: src/index.tsx [hunk 1]
```diff
@@ -669,17 +696,26 @@ class DraggableFlatList<T> extends React.Component<Props<T>, State> {
...
+                and(
+                  this.isScrolledDown,
+                  greaterOrEq(this.targetScrollOffset, this.scrollOffset)
+                )
               )
             ),
             [
+              // Finish scrolling
               set(this.isAutoscrolling.native, 0),
-              this.checkAutoscroll,
```
Comment (computerjazz) [inline src/index.tsx:682]: why was this removed? was it causing a bug or simply not necessary?

**Comment by computerjazz** [comment]
Thanks for the detailed bug report and pr!
Your changes look good -- I'll merge as soon as I test cross-platform. As I'm sure you've realized this project has required a lot of trial and error to get around fickle scrolling/performance behavior so even seemingly-minor changes carry some risk.

re: typescript build,  there's a few reasons I choose to release a built version
- It allows users to `skipLibCheck: true` so any draggable flatlist type errors don't affect users' own builds
- Some are interested in getting this working with react-native-web

File: src/index.tsx (hunk 1 above)
Comment (hsource) [reply src/index.tsx:682]: From what I understand, this was used before when `autoscroll` was a recursive function and resulted in duplicate calls to `autoscroll`. It can be kept in too to minimize changes!

The past code:

- `onScroll` would be called, and both call `this.checkAutoscroll` and resolve `this.onAutoscrollComplete`
  - `this.checkAutoscroll` would call `this.autoscroll` with `this.autoscrollParams`
  - `onAutoscrollComplete` would be called with `this.autoscrollParams`, which would resolve the promise that line 657 is awaiting `const nextScrollParams = await this.scrollToAsync(targetOffset);`
  - The end result is that on line 658, `this.autoscroll` would be called again with `this.autoscrollParams`

The end result is we just remove this duplicate call to `this.autoscroll`.

**Comment by hsource** [comment]
> Thanks for the detailed bug report and pr!
> Your changes look good -- I'll merge as soon as I test cross-platform. As I'm sure you've realized this project has required a lot of trial and error to get around fickle scrolling/performance behavior so even seemingly-minor changes carry some risk.
> 
> re: typescript build, there's a few reasons I choose to release a built version
> 
> * It allows users to `skipLibCheck: true` so any draggable flatlist type errors don't affect users' own builds
> * Some are interested in getting this working with react-native-web

That makes sense! Have you thought about including an example app inside this? I've pushed changes to react-native-cameraroll and react-native-webview before.

Let me test with my codebase on Android right now.

**Comment by computerjazz** [comment]
just tested android, looks good, thanks!
